File: booster-commons/src/test/java/io/github/booster/commons/retry/RetryConfigTest.java
package io.github.booster.commons.retry;

import io.github.booster.commons.metrics.MetricsRegistry;
import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.hamcrest.core.IsNull.notNullValue;
import static org.junit.jupiter.api.Assertions.assertThrows;

class RetryConfigTest {

    @Test
    void shouldCreateConfig() {
        assertThat(new RetryConfig(), notNullValue());
        assertThat(new RetryConfig(Map.of()), notNullValue());
        assertThat(new RetryConfig(Map.of("test", new RetrySetting())), notNullValue());
    }

    @Test
    void shouldNotCreateRetry() {
        assertThat(new RetryConfig().getOption("test").isDefined(), equalTo(false));
        assertThat(new RetryConfig(Map.of()).getOption("test").isDefined(), equalTo(false));
        assertThat(
                new RetryConfig(Map.of("abc", new RetrySetting())).getOption("test").isDefined(),
                equalTo(false)
        );
    }

    @Test
    void shouldCreateRetry() {
        RetrySetting setting = new RetrySetting();
        setting.setMaxAttempts(1);

        assertThat(
                new RetryConfig(Map.of("test", setting)).getOption("test").isDefined(),
                equalTo(true)
        );
    }

    @Test
    void shouldHandleRegistry() {
        RetrySetting setting = new RetrySetting();
        setting.setMaxAttempts(1);
        RetryConfig config = new RetryConfig(Map.of("test", setting));

        config.setMetricsRegistry(null);
        assertThat(
                config.getOption("test").isDefined(), equalTo(true)
        );
        assertThat(
                config.getOption("abc").isDefined(), equalTo(false)
        );

        config.setMetricsRegistry(new MetricsRegistry());
        assertThat(
                config.getOption("test").isDefined(), equalTo(true)
        );
        assertThat(
                config.getOption("abc").isDefined(), equalTo(false)
        );

        config.setMetricsRegistry(new MetricsRegistry(new SimpleMeterRegistry()));
        assertThat(
                config.getOption("test").isDefined(), equalTo(true)
        );
        assertThat(
                config.getOption("abc").isDefined(), equalTo(false)
        );
    }
}

File: booster-commons/src/test/kotlin/io/github/booster/commons/retry/RetryConfigTest.kt
package io.github.booster.commons.retry

import io.github.booster.commons.metrics.MetricsRegistry
import io.micrometer.core.instrument.simple.SimpleMeterRegistry
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.Test

internal class RetryConfigTest {
    @Test
    fun shouldCreateConfig() {
        assertThat(RetryConfig(), notNullValue())
        assertThat(RetryConfig(mapOf()), notNullValue())
        assertThat(RetryConfig(mapOf(Pair("test", RetrySetting()))), notNullValue())
    }

    @Test
    fun shouldNotCreateRetry() {
        assertThat(RetryConfig().tryGet("test").isDefined(), equalTo(false))
        assertThat(RetryConfig(mapOf()).tryGet("test").isDefined(), equalTo(false))
        assertThat(
            RetryConfig(mapOf(Pair("abc", RetrySetting()))).tryGet("test").isDefined(),
            equalTo(false)
        )
    }

    @Test
    fun shouldCreateRetry() {
        val setting = RetrySetting()
        setting.maxAttempts = 1
        assertThat(
            RetryConfig(mapOf(Pair("test", setting))).tryGet("test").isDefined(),
            equalTo(true)
        )
    }

    @Test
    fun `should allow null setting`() {
        val config = RetryConfig()
        config.setSettings(null)

        assertThat(
            config.tryGet("name").isDefined(),
            equalTo(false)
        )
    }

    @Test
    fun `should allow non-null setting`() {
        val setting = RetrySetting()
        setting.maxAttempts = 1

        val config = RetryConfig()
        config.setSettings(mapOf(Pair("test", setting)))

        assertThat(
            config.tryGet("test").isDefined(),
            equalTo(true)
        )
    }

    @Test
    fun shouldHandleRegistry() {
        val setting = RetrySetting()
        setting.maxAttempts = 1
        val config = RetryConfig(mapOf(Pair("test", setting)))
        config.setMetricsRegistry(null)
        assertThat(
            config.tryGet("test").isDefined(), equalTo(true)
        )
        assertThat(
            config.tryGet("abc").isDefined(), equalTo(false)
        )
        config.setMetricsRegistry(MetricsRegistry())
        assertThat(
            config.tryGet("test").isDefined(), equalTo(true)
        )
        assertThat(
            config.tryGet("abc").isDefined(), equalTo(false)
        )
        config.setMetricsRegistry(MetricsRegistry(SimpleMeterRegistry()))
        assertThat(
            config.tryGet("test").isDefined(), equalTo(true)
        )
        assertThat(
            config.tryGet("abc").isDefined(), equalTo(false)
        )
    }
}
